pilot/pkg/proxy/envoy/v2/xds.go

```diff
@@ -24,75 +27,135 @@ import (
 	out := &ConfigCache{
-		services: services,
-		configs:  configs,
+		serviceSuffix: "svc." + domainSuffix,
+		services:      services,
+		configs:       configs,
+		nodes:         make(map[string]time.Time),
 	}
-	out.snapshots = cache.NewSnapshotCache(false, out, nil)
+	out.snapshots = cache.NewSnapshotCache(false /* non-ADS mode */, out, nil /* add spam logger here */)
 	out.server = server.NewServer(out.snapshots, out)
 	return out
 }
 
-// Register with gRPC server
-func (cache *ConfigCache) Register(grpcServer *grpc.Server) {
-	v2.RegisterRouteDiscoveryServiceServer(grpcServer, cache.server)
-	// v2.RegisterClusterDiscoveryServiceServer(grpcServer, cache.server)
-	// v2.RegisterListenerDiscoveryServiceServer(grpcServer, cache.server)
+// RegisterOutput registers with a gRPC server
+// TODO: make sure the inputs are initialized before serving
+func (xds *ConfigCache) RegisterOutput(grpcServer *grpc.Server) {
+	v2.RegisterRouteDiscoveryServiceServer(grpcServer, xds.server)
 }
 
 // RegisterInput connects with the controllers
-func (cache *ConfigCache) RegisterInput(services model.Controller, configs model.ConfigStoreCache) {
-	if err := services.AppendServiceHandler(cache.OnServiceEvent); err != nil {
+// The route generation code needs to subscribe to:
+// - collection of services
+// - collection of virtual services
+func (xds *ConfigCache) RegisterInput(services model.Controller, configs model.ConfigStoreCache) {
+	if err := services.AppendServiceHandler(xds.OnServiceEvent); err != nil {
 		log.Fatal(err)
 	}
-	configs.RegisterEventHandler(model.VirtualService.Type, cache.OnConfigEvent)
-	configs.RegisterEventHandler(model.DestinationRule.Type, cache.OnConfigEvent)
+	configs.RegisterEventHandler(model.VirtualService.Type, xds.OnConfigEvent)
 }
 
-// ID ...
-func (cache *ConfigCache) ID(node *core.Node) string {
+// ID uses string-encoded node ID
+func (xds *ConfigCache) ID(node *core.Node) string {
 	return node.GetId()
 }
 
 // OnServiceEvent ...
-func (cache *ConfigCache) OnServiceEvent(svc *model.Service, event model.Event) {
-	// TODO
+func (xds *ConfigCache) OnServiceEvent(svc *model.Service, event model.Event) {
+	xds.Update()
 }
 
 // OnConfigEvent ...
-func (cache *ConfigCache) OnConfigEvent(svc model.Config, event model.Event) {
-	// TODO
+func (xds *ConfigCache) OnConfigEvent(svc model.Config, event model.Event) {
+	xds.Update()
 }
 
-// OnStreamOpen ...
-func (cache *ConfigCache) OnStreamOpen(int64, string) {}
+// Update performs the global re-compute
+// TODO make this more incremental
+func (xds *ConfigCache) Update() {
+	xds.mu.Lock()
+	defer xds.mu.Unlock()
+
+	xds.version++
+	services, _ := xds.services.Services()
+	virtualServices, _ := xds.configs.List(model.VirtualService.Type, model.NamespaceAll)
+	xds.hosts = v1alpha3.TranslateVirtualHosts(virtualServices, v1alpha3.IndexServices(services), xds.serviceSuffix)
```

Comment:
this entire piece should be in v1alpha3.. v1alpha3.BuildRDSConfig(), and any merging with services, virtual services, etc. is all v1alpha3's responsibility.